packages/liveblocks-core/src/comments/comment-body.ts

```diff
@@ -632,3 +632,112 @@ export async function stringifyCommentBody(
 
   return blocks.join(separator);
 }
+
+export type TransformCommentBodyOptions<U extends BaseUserMeta = DU> = {
+  /**
+   * Which format to transform the comment body to.
+   */
+  format?: "html" | "json";
+  /**
+   * A function that returns info from user IDs.
+   */
+  resolveUsers?: (
+    args: ResolveUsersArgs
+  ) => OptionalPromise<(U["info"] | undefined)[] | undefined>;
+};
+
+export type CommentBodyJsonMention = {
+  type: "mention";
+  user: string;
+};
+export type CommentBodyJsonLink = CommentBodyLink;
+export type CommentBodyJsonText = CommentBodyText;
+
+export type CommentBodyJsonInlineElement =
+  | CommentBodyJsonMention
+  | CommentBodyJsonLink
+  | CommentBodyJsonText;
+export type CommentBodyJsonParagraph = {
+  type: "paragraph";
+  children: CommentBodyJsonInlineElement[];
+};
+export type CommentBodyJsonBlockElement = CommentBodyJsonParagraph;
+/**
+ * A `CommentBodyJson` represents the content of a `CommentBody`
+ * with resolved users or not and with absolute urls
+ */
+export type CommentBodyJson = {
+  version: 1;
+  content: CommentBodyJsonBlockElement[];
+};
```

Comment:
See my answer [above](https://github.com/liveblocks/liveblocks/pull/1919#discussion_r1761080936). I do no think we should have the word `Resolved`. Or if you think we should force developers to resolve users 🤔

Reply:
Yeah maybe not `resolved`, I don't have other ideas immediately especially as the 2 formats are so similar, maybe it's a sign that we might not need 2 formats 🤔

Reply:
Yes this is a good question here 🤔 The only different thing here is the mention. So maybe it will more worth to extend the type `CommentBodyMention` to give him an extra optional key `user?: string` and fill it with a new function replacing `transformCommentBody` named `resolveCommentBody` running on each comments to fill the new field `user` with the resolver? 

WDYT of this new approach?

Reply:
I don't think we should add an optional field to the default type as it will never exist in the other cases, but I think this might be pushing us in the direction where we don't resolve before but only where the mention is used, imo this means that we would make a JSX version of `stringifyCommentBody` (not sure where we would ship this given the need to have `react` as a peer dep) where mentions are converted to `<span>@{user}</span>` without an intermediate JSON duplicate step

Reply:
I'm not sure to well understand you by making a JSX version of `stringifyCommentBody`. I do not see the point here.

From what I understand based on your feedback we should never touch the `CommentBody` and let users use it as it is and then expose the function `resolveUsersInCommentBody` to developers do whatever they want once they get the thread notification data.

However it's for developers an extra step as they already pass in params of `getThreadNotificationUnreadComments` (I put a comment btw about a new name) the option `resolveUsers` and do not feels like a good DX.

In the end here it look likes a dead end no? 😞

Reply:
My main issue is that the JSON path is only for React/JSX usage (if you want HTML, Markdown, or anything else, you'll use the string path) but right now it only adds the formatted mention with a resolver and format the links, and after that you still need to render it, so I wonder if we could skip the JSON step.

This React/JSX version of `stringifyCommentBody` could (very roughly) look like:
```tsx
const formattedBodyInJsx = transformCommentBody(body, {
  resolveUsers,
  components: {
    // By default it would render `<span>@{}</span>` but you can use this render prop to change what's displayed or use something else than a `span`, like `Text` from `react-email`
    mention: ({ element, user }) => {
        // `user` would be the resolved user if you passed something to `resolveUsers`
      	return <span>@{user}</span>
    }
  }
})
```

Here's what `stringifyCommentBody` does for mentions, very similar but as as string
![](https://github.com/user-attachments/assets/f0b0ef35-0289-4f92-ae7c-340b9a3f2b7b)

Reply:
No this is not what we discussed about before. The `JSON` path is not only for `React/JSX` usage.
The `JSON` path is for also other email templating engines like handlebars or liquid. 

And we were agreed that we didn't want to support JSX in this first milestone. 

We want only to offer either the `HTML` path were developers can take the HTML safe string we provide as it is if they want just to inject directly the HTML and the `JSON` path is when developers needs to use it in a whatever email templating engine.

Reply:
My assumption was that for Handlebars or Liquid you would use HTML as you don't need to insert content in comment bodies

Given that `CommentBody` is already JSON and usable, I think this door is still open for other setups, but making a second JSON format which only changes mentions and you still need to render paragraphs, links, italic, bold, etc yourself, I think that's not different/useful enough compared to using `CommentBody` directly

Reply:
That's why I proposed to add in `CommentBodyMention` an optional `user` key and fill it with `resolveUsersInCommentBody`. 

But if you think we should add it because it will never exist in other other cases, then my question is what can we do then besides duplicating?

Reply:
That's what I meant when I said that the current shape of `getThreadNotificationUnreadComments` (especially the fact that it currently asks for a resolver and you want to re-use it) is forcing us to think of only one option: outputting a transformed/resolved JSON

My preference is to match the front-end/`stringifyCommentBody` APIs where the resolving is done at the same time as the final format, so no need for an in-between JSON model that's resolved and so the question above doesn't exist. I'm not saying that means the JSX API needs to exist now, but I'm saying that we shouldn't do the resolving if we can't do the final format at the same time

Reply:
Sorry it's very hard for me to well understand your thoughts with just written words.

If I understand you well what's the point now to add a set of functions to help developers build an email without doing the final format?

The main goal here was to have utility functions to make developers build emails more easily based on a format we provide to them with resolved data like the author, users in comment body and the comment url. 

I do no think matching the front-end APIs here is really worth because we're not in a front-end context. We're in back-end context here. So we cannot really manipulate data in the same way than we will do in the front. 

So I would tend to just return the comment body as it is in then in the final format, no HTML nor markdown output but I strongly think we need to resolve mentioned users for sure.

Reply:
> ```tsx
> const formattedBodyInJsx = transformCommentBody(body, {
>   resolveUsers,
>   components: {
>     // By default it would render `<span>@{}</span>` but you can use this render prop to change what's displayed or use something else than a `span`, like `Text` from `react-email`
>     mention: ({ element, user }) => {
>         // `user` would be the resolved user if you passed something to `resolveUsers`
>         return <span>@{user}</span>
>     }
>   }
> })
> ```

Posting it here for posterity, we had a chat with @SugarDarius today and when I posted this idea above I forgot about my actual ideal API: since we're dealing with JSX anyway we could embrace React and have a component-based API (like in the react-ui package):

```tsx
function MyEmail() {
  const myUnreadComment = { ... }

  return (
    <div>
      {/* We offer this component, which works like `stringifyCommentBody` but with a React-friendly API like the `<Comment.Body />` primitive in the react-ui package  */}
  	  <CommentBody
  	    body={myUnreadComment.body}
        resolveUsers={...}
        components={{
          mention: ({ element, user }) => {
            // `user` would be the resolved user if you passed something to `resolveUsers`
            return <span>@{user}</span>
          }
        }}
      />
    </div>
  )
}
```